src/common/getType.ts

```diff
@@ -23,202 +37,474 @@ export function getType(expr: Expr, program: Program): ValueType {
-      }
-      return collectionType.member;
+      expectType(integerType(), integerType());
+      return textType(); // TODO narrow this
+    // unary
+    case "abs": {
+      expectType(integerType());
+      const t = types[0] as IntegerType;
+      return integerType(
+        0,
+        t.low === undefined || t.high === undefined
+          ? undefined
+          : -t.low > t.high
+          ? -t.low
+          : t.high
+      );
     }
-    case "array_get": {
-      const collectionType = arg0Type();
-      if (collectionType.type !== "Array") {
-        throw new Error("array_get must be used on an array");
-      }
-      return collectionType.member;
+    case "bit_not":
+      expectType(integerType());
+      return integerType();
+    case "neg": {
+      expectType(integerType());
+      const t = types[0] as IntegerType;
+      return integerType(
+        t.high === undefined ? undefined : -t.high,
+        t.low === undefined ? undefined : -t.low
+      );
     }
-    case "table_get": {
-      const collectionType = arg0Type();
-      if (collectionType.type !== "Table") {
-        throw new Error("table_get must be used on a table");
-      }
-      return collectionType.value;
+    case "not":
+      expectType(booleanType);
+      return booleanType;
+    case "int_to_text":
+    case "int_to_bin":
+    case "int_to_hex":
+      expectType(integerType());
+      return textType(); // TODO narrow this
+    case "text_to_int":
+      expectType(textType()); // TODO narrow this
+      return integerType();
+    case "bool_to_int":
+      expectType(booleanType);
+      return integerType(0, 1);
+    case "byte_to_char":
+      expectType(integerType(0, 255));
+      return integerType(0, 1);
+    case "list_length":
+      expectGenericType("List");
+      return integerType(0);
+    case "text_length":
+      expectType(textType());
+      return integerType(0, (types[0] as TextType).capacity);
+    case "text_split_whitespace":
+      expectType(textType());
+      return listType(types[0]);
+    case "sorted":
+      return listType(expectGenericType("List")[0]);
+    case "join":
+      expectType(listType(textType()));
+      return textType();
+    case "text_reversed":
+      expectType(textType());
+      return textType();
+    // other
+    case "true":
+      return booleanType;
+    case "false":
+      return booleanType;
+    case "argv":
+      return listType(textType());
+    case "print":
+    case "println":
+      return voidType;
+    case "text_replace":
+      expectType(textType(), textType(), textType());
+      return textType(); // TODO narrow this
+    case "text_get_slice": {
+      expectType(textType(), integerType(0), integerType(0));
+      const [t, i1, i2] = types as [TextType, IntegerType, IntegerType];
+      return textType(
+        t.capacity === undefined && i2.high === undefined
+          ? undefined
+          : t.capacity === undefined
+          ? Number(i2.high! - i1.low!)
```

Comment:
Where does `i1.low` get checked to let you conclude it's a number?

Reply:
It doesn't thank you for being thorough with the reviews. :)